### ISSUE TITLE
ivy-immediate-done with alist candidates

### ISSUE THREAD

**Issue by justbur** [abo-abo/swiper#878]
This was just accidental on my part, but if you use `ivy-immediate-done` when the candidates are an alist a string gets passed to the action which expects a cons cell. Should this be handled in the action or in `ivy-immediate-done`?

**Comment by abo-abo** [comment]
Can't reproduce with this:

```elisp
(defun test-action (x)
  (message "test-action: %S" x))

(ivy-read "test:" '(("a" . 1) ("b" . 2))
          :action 'test-action)
```

Results in `test-action: (#("a" 0 1 (idx 0)) . 1)` both for <kbd>C-j</kbd> and <kbd>C-M-j</kbd>.

**Comment by justbur** [comment]
Sorry, I was referring to the case when the key is not present in the alist. With your example try `c C-M-j`. You should get `test-action: "c"` which would be an error if the action didn't handle strings.

**Comment by justbur** [comment]
I was just thinking that `:require-match t` would be appropriate here, but that doesn't seem to enforce anything with `C-M-j`

**Comment by basil-conto** [comment]
> With your example try `c C-M-j`. You should get `test-action: "c"` which would be an error if the action didn't handle strings.

Right, but what can be done about this, other documenting it and fixing any existing naive action functions?

> I was just thinking that `:require-match t` would be appropriate here, but that doesn't seem to enforce anything with `C-M-j`

I haven't tested this particular example, but there are known quirks with `:require-match`, e.g. #1631, #1679.
